### PR TITLE
`Development`: Increase client test coverage of category selector

### PR DESCRIPTION
**Pull request by filip-gregurevic**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

We want to increase client test coverage.

### Description

This PR increases client test coverage for app/shared/category-selector.

### Steps for Testing

Execute the client tests for  app/shared/category-selector

#### Code Review
- [x] Review 1
- [x] Review 2

### Test Coverage

category-selector.component.ts:  73.68 | 97.36

### CODE REVIEW

File: src/main/webapp/app/shared/category-selector/category-selector.component.ts [hunk 1]
```diff
@@ -52,22 +50,22 @@ export class CategorySelectorComponent implements OnInit, OnChanges {
         );
     }
 
-    private categoriesAsStringArray() {
```
Comment (julian-christl) [inline src/main/webapp/app/shared/category-selector/category-selector.component.ts:55]: I don't like setting all methods to public. You don't need to test every method on its own to test all lines and branches.
Please revert the changes to the access modifiers and focus on testing the already public methods. You can think about what scenarios reach the respective part of the code you are trying to test and use that as a test case.

Keep in mind, you can also modify `test:diff` by adding a `--coverage` parameter to it and you can quickly check what lines of your component are not yet covered. Just don't commit the change, we don't need it on default.

File: src/main/webapp/app/shared/category-selector/category-selector.component.ts [hunk 2]
```diff
@@ -52,22 +50,22 @@ export class CategorySelectorComponent implements OnInit, OnChanges {
...
+    categoriesAsStringArray(): string[] {
```
Comment (julian-christl) [inline src/main/webapp/app/shared/category-selector/category-selector.component.ts:53]: I don't like setting all methods to public. You don't need to test every method on its own to test all lines and branches.
Please revert the changes to the access modifiers and focus on testing the already public methods. You can think about what scenarios reach the respective part of the code you are trying to test and use that as a test case.

Keep in mind, you can also modify test:diff by adding a --coverage parameter to it and you can quickly check what lines of your component are not yet covered. Just don't commit the change, we don't need it on default.

**Review by julian-christl — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/category-selector/category-selector.component.ts (hunk 2 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/shared/category-selector/category-selector.component.ts]: @julian-christl Good call! I changed it now and thanks for the good hints 👍🏼

**Review by Malyuk-A — DISMISSED** [review]
Changes look reasonable to me.

File: src/main/webapp/app/shared/category-selector/category-selector.component.ts [hunk 3]
```diff
@@ -52,18 +50,18 @@ export class CategorySelectorComponent implements OnInit, OnChanges {
...
-    private existingCategoriesAsStringArray() {
+    private existingCategoriesAsStringArray(): string[] {
         if (!this.existingCategories) {
             return [];
         }
-        return this.existingCategories.map((exerciseCategory) => exerciseCategory.category?.toLowerCase());
+        return this.existingCategories.map((exerciseCategory) => exerciseCategory.category!.toLowerCase());
     }
 
     // if the user types in something, we need to filter for the matching categories
```
Comment (julian-christl) [inline src/main/webapp/app/shared/category-selector/category-selector.component.ts:67]: According to the coverage report, this method is not tested at all. Could you add tests that include the user input and the resulting filter?

**Review by julian-christl — COMMENTED** [review]
Also, your coverage numbers are 1) probably switched and 2) not up to date

File: src/main/webapp/app/shared/category-selector/category-selector.component.ts (hunk 3 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/shared/category-selector/category-selector.component.ts:67]: @julian-christl Good catch, I added a test for this case now 👍🏼 
Also, I updated the percentages (They're in the order as described in the PR template statements | branches )

**Review by julian-christl — APPROVED** [review]
Looks good, thanks for the changes!

**Review by canberkanar — COMMENTED** [review]
LGTM 👍🏻

**Review by canberkanar — APPROVED** [review]

File: src/test/javascript/spec/component/shared/category-selector.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    });
+
+    it('should not create duplicate item on add', () => {
+        comp.categories = [category6];
+        const event = { value: 'category6', chipInput: { clear: () => {} } as MatChipInput } as MatChipInputEvent;
+        fixture.detectChanges();
+        comp.onItemAdd(event);
+
+        expect(comp.categories).toEqual([category6]);
+        expect(emitSpy).not.toHaveBeenCalled();
```
Comment (willergeorg) [inline src/test/javascript/spec/component/shared/category-selector.component.spec.ts:146]:
```suggestion
        expect(emitSpy).not.toHaveBeenCalledTimes(1);
```

**Review by willergeorg — APPROVED** [review]
I only found two nits. Besides that, lgtm!

**Review by julian-christl — APPROVED** [review]

**Review by Hialus — APPROVED** [review]
Code changes look good to me. 👍🏻

**Review by canberkanar — APPROVED** [review]

**Review by willergeorg — APPROVED** [review]

**Pull request merged**